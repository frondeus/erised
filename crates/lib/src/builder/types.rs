use std::sync::Arc;

use crate::heap_types::*;
use rustdoc_types::Id;

use super::{Builder, Cache, Error, Result};

impl Builder {
    pub(crate) fn build_type(&self, cache: &mut Cache, ty: &rustdoc_types::Type) -> Result<Type> {
        match ty {
            rustdoc_types::Type::ResolvedPath(path) => {
                Ok(Type::ResolvedPath(self.build_path(cache, path)?))
            }
            rustdoc_types::Type::DynTrait(_) => todo!(),
            rustdoc_types::Type::Generic(_) => todo!(),
            rustdoc_types::Type::Primitive(_) => todo!(),
            rustdoc_types::Type::FunctionPointer(_) => todo!(),
            rustdoc_types::Type::Tuple(_) => todo!(),
            rustdoc_types::Type::Slice(_) => todo!(),
            rustdoc_types::Type::Array { type_, len } => todo!(),
            rustdoc_types::Type::ImplTrait(generic_bounds) => {
                let mut imp = vec![];
                for bound in generic_bounds {
                    imp.push(self.build_generic_bound(cache, bound)?);
                }
                Ok(Type::ImplTrait(imp))
            }
            rustdoc_types::Type::Infer => todo!(),
            rustdoc_types::Type::RawPointer { mutable, type_ } => todo!(),
            rustdoc_types::Type::BorrowedRef {
                lifetime,
                mutable,
                type_,
            } => todo!(),
            rustdoc_types::Type::QualifiedPath {
                name,
                args,
                self_type,
                trait_,
            } => todo!(),
        }
    }

    pub(crate) fn build_generic_bound(
        &self,
        cache: &mut Cache,
        bound: &rustdoc_types::GenericBound,
    ) -> Result<GenericBound> {
        Ok(match bound {
            rustdoc_types::GenericBound::TraitBound {
                trait_,
                generic_params: source_generic_params,
                modifier,
            } => {
                let mut generic_params = vec![];
                for generic_param in source_generic_params {
                    generic_params.push(self.build_generic_param_def(cache, generic_param)?);
                }
                GenericBound::TraitBound {
                    trait_: self.build_path(cache, trait_)?,
                    generic_params,
                    modifier: self.build_trait_bound_modifier(cache, modifier)?,
                }
            }
            rustdoc_types::GenericBound::Outlives(_) => todo!(),
        })
    }

    pub(crate) fn build_trait_bound_modifier(
        &self,
        cache: &mut Cache,
        modifier: &rustdoc_types::TraitBoundModifier,
    ) -> Result<TraitBoundModifier> {
        Ok(match modifier {
            rustdoc_types::TraitBoundModifier::None => TraitBoundModifier::None,
            rustdoc_types::TraitBoundModifier::Maybe => TraitBoundModifier::Maybe,
            rustdoc_types::TraitBoundModifier::MaybeConst => TraitBoundModifier::MaybeConst,
        })
    }

    pub(crate) fn build_generic_param_def(
        &self,
        cache: &mut Cache,
        source: &rustdoc_types::GenericParamDef,
    ) -> Result<GenericParamDef> {
        Ok(GenericParamDef {
            name: source.name.clone(),
            kind: self.build_generic_param_def_kind(cache, &source.kind)?,
        })
    }

    pub(crate) fn build_generic_param_def_kind(
        &self,
        cache: &mut Cache,
        source: &rustdoc_types::GenericParamDefKind,
    ) -> Result<GenericParamDefKind> {
        Ok(match source {
            rustdoc_types::GenericParamDefKind::Lifetime { outlives } => {
                GenericParamDefKind::Lifetime {
                    outlives: outlives.clone(),
                }
            }
            rustdoc_types::GenericParamDefKind::Type {
                bounds: source_bounds,
                default,
                synthetic,
            } => {
                let mut bounds = vec![];
                for bound in source_bounds {
                    bounds.push(self.build_generic_bound(cache, bound)?);
                }
                GenericParamDefKind::Type {
                    bounds,
                    default: match default.as_ref() {
                        Some(default) => Some(self.build_type(cache, default)?),
                        None => None,
                    },
                    synthetic: *synthetic,
                }
            }
            rustdoc_types::GenericParamDefKind::Const { type_, default } => todo!(),
        })
    }
}
