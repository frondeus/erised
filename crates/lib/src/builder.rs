use std::{
    collections::HashMap,
    sync::{Arc, Weak},
};

use crate::heap_types::*;
use rustdoc_types::Id;
use thiserror::Error;

pub use rustdoc_json::PackageTarget;

#[derive(Debug, Error)]
pub enum Error {
    #[error(transparent)]
    BuildError(#[from] rustdoc_json::BuildError),

    #[error(transparent)]
    IO(#[from] std::io::Error),

    #[error(transparent)]
    JSON(#[from] serde_json::Error),

    #[error("Could not find item with id: {0:?}")]
    CouldNotFind(rustdoc_types::Id),

    #[error("Could not find crate with id: {0:?}")]
    CouldNotFindCrate(u32),
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

pub struct Builder {
    source: rustdoc_types::Crate,
    // reflect_id: rustdoc_types::Id,
    root: rustdoc_types::Item,
}

#[derive(Default)]
pub(crate) struct Cache {
    weak_items: HashMap<Id, Weak<Item>>,
    items: HashMap<Id, Arc<Item>>,
    summaries: HashMap<Id, Arc<ItemSummary>>,
    crates: HashMap<u32, Arc<ExternalCrate>>,
}

mod crates;
mod generics;
mod items;
mod paths;
mod types;

impl Builder {
    pub fn load(
        manifest_dir: impl AsRef<std::path::Path>,
        options: impl Fn(rustdoc_json::Builder) -> rustdoc_json::Builder,
    ) -> Result<Option<Crate>> {
        let manifest_dir = manifest_dir.as_ref();

        let builder = rustdoc_json::Builder::default()
            .toolchain("nightly")
            .manifest_path(manifest_dir.join("Cargo.toml"))
            .target_dir(manifest_dir.join("erised_target"))
            .document_private_items(true);

        let json_path = options(builder).build()?;
        let file = std::fs::OpenOptions::new().read(true).open(&json_path)?;

        let source: rustdoc_types::Crate = serde_json::from_reader(file)?;

        // let reflect_id = source
        //     .paths
        //     .iter()
        //     .find(|(_path, val)| val.path == &["erised", "ToReflect"])
        //     .map(|(p, _)| p)
        //     .cloned();

        // let reflect_id = match reflect_id {
        //     Some(id) => id,
        //     None => {
        //         eprintln!(
        //             "cargo-warning=Reflection crate did not find a single usage of erised::ToReflect"
        //         );
        //         eprintln!("cargo-warning=It tried to load {json_path:?}");
        //         return Ok(None);
        //     }
        // };

        let root = source
            .index
            .get(&source.root)
            .ok_or_else(|| Error::CouldNotFind(source.root.clone()))?
            .clone();

        let builder = Builder {
            source,
            // reflect_id,
            root,
        };

        builder.build().map(Some)
    }

    pub fn build(self) -> Result<Crate> {
        let mut cache = Default::default();
        let root = (*self
            .get_item(&mut cache, &self.source.root)?
            .upgrade()
            .expect("Module"))
        .clone()
        .as_module()
        .expect("Module")
        .into();

        Ok(Crate {
            root,
            crate_version: self.source.crate_version,
            all_items: cache.items.into_iter().map(|(_k, v)| v).collect(),
            summaries: vec![],
            external_crates: vec![],
        })
    }
}
