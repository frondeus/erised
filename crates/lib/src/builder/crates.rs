use std::sync::Arc;

use crate::heap_types::*;

use super::{Builder, Cache, Error, Result};

impl Builder {
    pub(crate) fn get_crate(&self, cache: &mut Cache, id: u32) -> Result<Arc<ExternalCrate>> {
        if let Some(cached) = cache.crates.get(&id) {
            return Ok(cached.clone());
        }

        if (id == self.root.crate_id) {
            let krate = Arc::new(ExternalCrate {
                name: self.root.name.clone().unwrap(),
                html_root_url: None,
            });
            cache.crates.insert(id, krate.clone());
            return Ok(krate);
        }

        let krate = self
            .source
            .external_crates
            .get(&id)
            .ok_or_else(|| Error::CouldNotFindCrate(id))?;

        let krate = Arc::new(ExternalCrate {
            name: krate.name.clone(),
            html_root_url: krate.html_root_url.clone(),
        });

        cache.crates.insert(id, krate.clone());

        Ok(krate)
    }
}
