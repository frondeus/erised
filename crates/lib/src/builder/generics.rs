use std::sync::Arc;

use crate::heap_types::*;
use rustdoc_types::Id;

use super::{Builder, Cache, Error, Result};

impl Builder {
    pub(crate) fn build_generics(
        &self,
        cache: &mut Cache,
        source: &rustdoc_types::Generics,
    ) -> Result<Generics> {
        let mut params = vec![];
        for param in &source.params {
            params.push(self.build_generic_param_def(cache, param)?);
        }

        let mut where_predicates = vec![];
        for where_predicate in &source.where_predicates {
            where_predicates.push(self.build_where_predicate(cache, where_predicate)?);
        }

        Ok(Generics {
            params,
            where_predicates,
        })
    }

    pub(crate) fn build_where_predicate(
        &self,
        cache: &mut Cache,
        source: &rustdoc_types::WherePredicate,
    ) -> Result<WherePredicate> {
        Ok(match source {
            rustdoc_types::WherePredicate::BoundPredicate {
                type_,
                bounds,
                generic_params,
            } => todo!(),
            rustdoc_types::WherePredicate::RegionPredicate { lifetime, bounds } => todo!(),
            rustdoc_types::WherePredicate::EqPredicate { lhs, rhs } => todo!(),
        })
    }
}
