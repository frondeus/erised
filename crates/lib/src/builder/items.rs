use std::sync::Arc;

use crate::heap_types::*;
use rustdoc_types::Id;

use super::{Builder, Cache, Error, Result};

mod deprecation;
mod spans;
mod visibility;

mod functions;
mod imports;
mod modules;

impl Builder {
    pub(crate) fn build_item_meta(
        &self,
        cache: &mut Cache,
        item: &rustdoc_types::Item,
    ) -> Result<ItemMeta> {
        Ok(ItemMeta {
            krate: self.get_crate(cache, item.crate_id)?,
            span: self.build_span(cache, item.span.as_ref())?,
            visibility: self.build_visibility(cache, &item.visibility)?,
            docs: item.docs.clone(),
            attrs: item.attrs.clone(),
            deprecation: self.build_deprecation(cache, item.deprecation.as_ref())?,
        })
    }

    pub(crate) fn get_item(&self, cache: &mut Cache, id: &Id) -> Result<Arc<Item>> {
        if let Some(cached) = cache.items.get(id) {
            return Ok(cached.clone());
        }

        let item = self
            .source
            .index
            .get(id)
            .ok_or_else(|| Error::CouldNotFind(id.clone()))?;

        let meta = self.build_item_meta(cache, item)?;
        let name = item.name.clone();
        let item = self.build_item(cache, name, meta, &item.inner)?;
        let item = Arc::new(item);

        cache.items.insert(id.clone(), item.clone());

        Ok(item)
    }

    pub(crate) fn build_item(
        &self,
        cache: &mut Cache,
        name: Option<String>,
        meta: ItemMeta,
        item: &rustdoc_types::ItemEnum,
    ) -> Result<Item> {
        match item {
            rustdoc_types::ItemEnum::Module(module) => {
                Ok(Item::Module(self.build_module(cache, name, meta, module)?))
            }
            rustdoc_types::ItemEnum::ExternCrate { name, rename } => todo!(),
            rustdoc_types::ItemEnum::Import(import) => {
                Ok(Item::Import(self.build_import(cache, meta, import)?))
            }
            rustdoc_types::ItemEnum::Union(_) => todo!(),
            rustdoc_types::ItemEnum::Struct(_) => todo!(),
            rustdoc_types::ItemEnum::StructField(_) => todo!(),
            rustdoc_types::ItemEnum::Enum(_) => todo!(),
            rustdoc_types::ItemEnum::Variant(_) => todo!(),
            rustdoc_types::ItemEnum::Function(func) => Ok(Item::Function(
                self.build_function(cache, name, meta, func)?,
            )),
            rustdoc_types::ItemEnum::Trait(_) => todo!(),
            rustdoc_types::ItemEnum::TraitAlias(_) => todo!(),
            rustdoc_types::ItemEnum::Impl(_) => todo!(),
            rustdoc_types::ItemEnum::Typedef(_) => todo!(),
            rustdoc_types::ItemEnum::OpaqueTy(_) => todo!(),
            rustdoc_types::ItemEnum::Constant(_) => todo!(),
            rustdoc_types::ItemEnum::Static(_) => todo!(),
            rustdoc_types::ItemEnum::ForeignType => todo!(),
            rustdoc_types::ItemEnum::Macro(_) => todo!(),
            rustdoc_types::ItemEnum::ProcMacro(_) => todo!(),
            rustdoc_types::ItemEnum::Primitive(_) => todo!(),
            rustdoc_types::ItemEnum::AssocConst { type_, default } => todo!(),
            rustdoc_types::ItemEnum::AssocType {
                generics,
                bounds,
                default,
            } => todo!(),
        }
    }

    pub(crate) fn get_item_summary(&self, cache: &mut Cache, id: &Id) -> Result<Arc<ItemSummary>> {
        if let Some(cached) = cache.summaries.get(id) {
            return Ok(cached.clone());
        }

        let item = self
            .source
            .paths
            .get(id)
            .ok_or_else(|| Error::CouldNotFind(id.clone()))?;

        let item = self.build_item_summary(cache, &item)?;
        let item = Arc::new(item);

        cache.summaries.insert(id.clone(), item.clone());

        Ok(item)
    }

    pub(crate) fn build_item_summary(
        &self,
        cache: &mut Cache,
        item: &rustdoc_types::ItemSummary,
    ) -> Result<ItemSummary> {
        Ok(ItemSummary {
            krate: self.get_crate(cache, item.crate_id)?,
            path: item.path.clone(),
            kind: match item.kind {
                rustdoc_types::ItemKind::Module => ItemKind::Module,
                rustdoc_types::ItemKind::ExternCrate => ItemKind::ExternCrate,
                rustdoc_types::ItemKind::Import => ItemKind::Import,
                rustdoc_types::ItemKind::Struct => ItemKind::Struct,
                rustdoc_types::ItemKind::StructField => ItemKind::StructField,
                rustdoc_types::ItemKind::Union => ItemKind::Union,
                rustdoc_types::ItemKind::Enum => ItemKind::Enum,
                rustdoc_types::ItemKind::Variant => ItemKind::Variant,
                rustdoc_types::ItemKind::Function => ItemKind::Function,
                rustdoc_types::ItemKind::Typedef => ItemKind::Typedef,
                rustdoc_types::ItemKind::OpaqueTy => ItemKind::OpaqueTy,
                rustdoc_types::ItemKind::Constant => ItemKind::Constant,
                rustdoc_types::ItemKind::Trait => ItemKind::Trait,
                rustdoc_types::ItemKind::TraitAlias => ItemKind::TraitAlias,
                rustdoc_types::ItemKind::Impl => ItemKind::Impl,
                rustdoc_types::ItemKind::Static => ItemKind::Static,
                rustdoc_types::ItemKind::ForeignType => ItemKind::ForeignType,
                rustdoc_types::ItemKind::Macro => ItemKind::Macro,
                rustdoc_types::ItemKind::ProcAttribute => ItemKind::ProcAttribute,
                rustdoc_types::ItemKind::ProcDerive => ItemKind::ProcDerive,
                rustdoc_types::ItemKind::AssocConst => ItemKind::AssocConst,
                rustdoc_types::ItemKind::AssocType => ItemKind::AssocType,
                rustdoc_types::ItemKind::Primitive => ItemKind::Primitive,
                rustdoc_types::ItemKind::Keyword => ItemKind::Keyword,
            },
        })
    }

    pub(crate) fn get_identifiable(&self, cache: &mut Cache, id: &Id) -> Result<Identifiable> {
        let item = self.get_item(cache, id);
        match item {
            Ok(item) => Ok(Identifiable::Item(item)),
            Err(Error::CouldNotFind(_)) => {
                let summary = self.get_item_summary(cache, id)?;
                Ok(Identifiable::Summary(summary))
            }
            Err(e) => Err(e),
        }
    }
}
