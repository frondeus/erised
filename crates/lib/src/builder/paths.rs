use std::sync::Arc;

use crate::heap_types::*;
use rustdoc_types::Id;

use super::{Builder, Cache, Error, Result};

impl Builder {
    pub(crate) fn build_path(&self, cache: &mut Cache, ty: &rustdoc_types::Path) -> Result<Path> {
        let split = ty.name.rsplit_once("::");
        let (prefix, name) = match split {
            Some((prefix, name)) => (prefix.to_owned(), name.to_owned()),
            None => ("".to_owned(), ty.name.clone()),
        };

        Ok(Path {
            name,
            prefix,
            target: self.get_identifiable(cache, &ty.id)?,
            args: match ty.args.as_ref() {
                None => None,
                Some(args) => Some(Box::new(self.build_generic_args(cache, &args)?)),
            },
        })
    }

    pub(crate) fn build_generic_args(
        &self,
        cache: &mut Cache,
        args: &rustdoc_types::GenericArgs,
    ) -> Result<GenericArgs> {
        match args {
            rustdoc_types::GenericArgs::AngleBracketed {
                args: source_args,
                bindings: source_bindings,
            } => {
                let mut args = vec![];
                for arg in source_args {
                    args.push(self.build_generic_arg(cache, arg)?);
                }
                let mut bindings = vec![];
                for binding in source_bindings {
                    bindings.push(self.build_type_binding(cache, binding)?);
                }
                Ok(GenericArgs::AngleBracketed { args, bindings })
            }
            rustdoc_types::GenericArgs::Parenthesized { inputs, output } => todo!(),
        }
    }

    pub(crate) fn build_generic_arg(
        &self,
        cache: &mut Cache,
        arg: &rustdoc_types::GenericArg,
    ) -> Result<GenericArg> {
        Ok(match arg {
            rustdoc_types::GenericArg::Lifetime(_) => todo!(),
            rustdoc_types::GenericArg::Type(ty) => GenericArg::Type(self.build_type(cache, ty)?),
            rustdoc_types::GenericArg::Const(_) => todo!(),
            rustdoc_types::GenericArg::Infer => todo!(),
        })
    }

    pub(crate) fn build_type_binding(
        &self,
        cache: &mut Cache,
        source: &rustdoc_types::TypeBinding,
    ) -> Result<TypeBinding> {
        Ok(TypeBinding {
            name: source.name.clone(),
            args: self.build_generic_args(cache, &source.args)?,
            binding: self.build_type_binding_kind(cache, &source.binding)?,
        })
    }

    pub(crate) fn build_type_binding_kind(
        &self,
        cache: &mut Cache,
        source: &rustdoc_types::TypeBindingKind,
    ) -> Result<TypeBindingKind> {
        Ok(match source {
            rustdoc_types::TypeBindingKind::Equality(term) => {
                TypeBindingKind::Equality(self.build_term(cache, &term)?)
            }
            rustdoc_types::TypeBindingKind::Constraint(generic_bounds) => {
                let mut bounds = vec![];
                for bound in generic_bounds {
                    bounds.push(self.build_generic_bound(cache, bound)?);
                }
                TypeBindingKind::Constraint(bounds)
            }
        })
    }

    pub(crate) fn build_term(
        &self,
        cache: &mut Cache,
        source: &rustdoc_types::Term,
    ) -> Result<Term> {
        Ok(match source {
            rustdoc_types::Term::Type(ty) => Term::Type(self.build_type(cache, ty)?),
            rustdoc_types::Term::Constant(c) => todo!(),
        })
    }
}
