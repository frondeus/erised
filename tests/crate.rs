use std::path::PathBuf;

use erised::builder::BuilderOpts;
use erised_tests::pretty_print;

#[test]
fn erised() {
    let manifest_dir = PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").unwrap_or_default());

    let krate = BuilderOpts::default()
        .manifest_dir(manifest_dir)
        .package("erised")
        .load()
        .expect("JSON info")
        .build()
        .expect("Rust info");

    let info = krate.generate_static();

    let info = pretty_print(info).expect("formatted");

    insta::assert_display_snapshot!(info);
}

#[test]
fn erised_tests() {
    let manifest_dir = PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").unwrap_or_default());

    let krate = BuilderOpts::default()
        .manifest_dir(manifest_dir)
        .package("erised")
        .load()
        .expect("JSON info")
        .build()
        .expect("Rust info");

    let info = krate.generate_static();

    let info = pretty_print(info).expect("formatted");

    insta::assert_display_snapshot!(info);
}
