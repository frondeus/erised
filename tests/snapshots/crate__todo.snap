---
source: tests/crate.rs
expression: info
---
fn test() {
    erised::types::Crate {
        root: || erised::types::Module {
            name: "erised_tests",
            meta: erised::types::ItemMeta {
                krate: || erised::types::ExternalCrate {
                    name: "erised_tests",
                    html_root_url: None,
                },
                span: Some(erised::types::Span {
                    filename: "src/lib.rs",
                    begin: (1usize, 0usize),
                    end: (60usize, 1usize),
                }),
                visibility: erised::types::Visibility::Public,
                docs: None,
                attrs: &[],
                deprecation: None,
            },
            is_crate: true,
            items: &[
                erised::types::Identifiable::Item(|| {
                    erised::types::Item::Function(erised::types::Function {
                        name: "test",
                        meta: erised::types::ItemMeta {
                            krate: || erised::types::ExternalCrate {
                                name: "erised_tests",
                                html_root_url: None,
                            },
                            span: Some(erised::types::Span {
                                filename: "src/lib.rs",
                                begin: (13usize, 0usize),
                                end: (13usize, 24usize),
                            }),
                            visibility: erised::types::Visibility::Public,
                            docs: None,
                            attrs: &[],
                            deprecation: None,
                        },
                        decl: erised::types::FnDecl {
                            inputs: &[erised::types::FnInput {
                                pat: "t",
                                ty: erised::types::Type::ResolvedPath(erised::types::Path {
                                    name: "Crate",
                                    prefix: "",
                                    target: erised::types::Identifiable::Summary(|| {
                                        erised::types::ItemSummary {
                                            krate: || erised::types::ExternalCrate {
                                                name: "erised",
                                                html_root_url: None,
                                            },
                                            path: &["erised", "heap_types", "Crate"],
                                            kind: erised::types::ItemKind::Struct,
                                        }
                                    }),
                                    args: Some(|| erised::types::GenericArgs::AngleBracketed {
                                        args: &[],
                                        bindings: &[],
                                    }),
                                }),
                            }],
                            output: None,
                            c_variadic: false,
                        },
                        generics: erised::types::Generics {
                            params: &[],
                            where_predicates: &[],
                        },
                        header: erised::types::Header {
                            const_: false,
                            unsafe_: false,
                            async_: false,
                            abi: erised::types::Abi::Rust,
                        },
                        has_body: true,
                    })
                }),
                erised::types::Identifiable::Item(|| {
                    erised::types::Item::Function(erised::types::Function {
                        name: "pretty_print_item",
                        meta: erised::types::ItemMeta {
                            krate: || erised::types::ExternalCrate {
                                name: "erised_tests",
                                html_root_url: None,
                            },
                            span: Some(erised::types::Span {
                                filename: "src/lib.rs",
                                begin: (15usize, 0usize),
                                end: (21usize, 1usize),
                            }),
                            visibility: erised::types::Visibility::Public,
                            docs: None,
                            attrs: &[],
                            deprecation: None,
                        },
                        decl: erised::types::FnDecl {
                            inputs: &[erised::types::FnInput {
                                pat: "tokens",
                                ty: erised::types::Type::ImplTrait(&[
                                    erised::types::GenericBound::TraitBound {
                                        trait_: erised::types::Path {
                                            name: "ToTokens",
                                            prefix: "",
                                            target: erised::types::Identifiable::Summary(|| {
                                                erised::types::ItemSummary {
                                                    krate: || erised::types::ExternalCrate {
                                                        name: "quote",
                                                        html_root_url: Some(
                                                            "https://docs.rs/quote/1.0.26/",
                                                        ),
                                                    },
                                                    path: &["quote", "to_tokens", "ToTokens"],
                                                    kind: erised::types::ItemKind::Trait,
                                                }
                                            }),
                                            args: Some(|| {
                                                erised::types::GenericArgs::AngleBracketed {
                                                    args: &[],
                                                    bindings: &[],
                                                }
                                            }),
                                        },
                                        generic_params: &[],
                                        modifier: erised::types::TraitBoundModifier::None,
                                    },
                                ]),
                            }],
                            output: Some(erised::types::Type::ResolvedPath(erised::types::Path {
                                name: "Result",
                                prefix: "",
                                target: erised::types::Identifiable::Summary(|| {
                                    erised::types::ItemSummary {
                                        krate: || erised::types::ExternalCrate {
                                            name: "core",
                                            html_root_url: Some(
                                                "https://doc.rust-lang.org/nightly/",
                                            ),
                                        },
                                        path: &["core", "result", "Result"],
                                        kind: erised::types::ItemKind::Enum,
                                    }
                                }),
                                args: Some(|| erised::types::GenericArgs::AngleBracketed {
                                    args: &[
                                        erised::types::GenericArg::Type(
                                            erised::types::Type::ResolvedPath(
                                                erised::types::Path {
                                                    name: "String",
                                                    prefix: "",
                                                    target: erised::types::Identifiable::Summary(
                                                        || erised::types::ItemSummary {
                                                            krate: || {
                                                                erised :: types :: ExternalCrate { name : "alloc" , html_root_url : Some ("https://doc.rust-lang.org/nightly/") }
                                                            },
                                                            path: &["alloc", "string", "String"],
                                                            kind: erised::types::ItemKind::Struct,
                                                        },
                                                    ),
                                                    args: Some(|| {
                                                        erised::types::GenericArgs::AngleBracketed {
                                                            args: &[],
                                                            bindings: &[],
                                                        }
                                                    }),
                                                },
                                            ),
                                        ),
                                        erised::types::GenericArg::Type(
                                            erised::types::Type::ResolvedPath(
                                                erised::types::Path {
                                                    name: "Error",
                                                    prefix: "",
                                                    target: erised::types::Identifiable::Summary(
                                                        || erised::types::ItemSummary {
                                                            krate: || {
                                                                erised :: types :: ExternalCrate { name : "anyhow" , html_root_url : Some ("https://docs.rs/anyhow/1.0.70/") }
                                                            },
                                                            path: &["anyhow", "Error"],
                                                            kind: erised::types::ItemKind::Struct,
                                                        },
                                                    ),
                                                    args: Some(|| {
                                                        erised::types::GenericArgs::AngleBracketed {
                                                            args: &[],
                                                            bindings: &[],
                                                        }
                                                    }),
                                                },
                                            ),
                                        ),
                                    ],
                                    bindings: &[],
                                }),
                            })),
                            c_variadic: false,
                        },
                        generics: erised::types::Generics {
                            params: &[erised::types::GenericParamDef {
                                name: "impl ToTokens",
                                kind: erised::types::GenericParamDefKind::Type {
                                    bounds: &[erised::types::GenericBound::TraitBound {
                                        trait_: erised::types::Path {
                                            name: "ToTokens",
                                            prefix: "",
                                            target: erised::types::Identifiable::Summary(|| {
                                                erised::types::ItemSummary {
                                                    krate: || erised::types::ExternalCrate {
                                                        name: "quote",
                                                        html_root_url: Some(
                                                            "https://docs.rs/quote/1.0.26/",
                                                        ),
                                                    },
                                                    path: &["quote", "to_tokens", "ToTokens"],
                                                    kind: erised::types::ItemKind::Trait,
                                                }
                                            }),
                                            args: Some(|| {
                                                erised::types::GenericArgs::AngleBracketed {
                                                    args: &[],
                                                    bindings: &[],
                                                }
                                            }),
                                        },
                                        generic_params: &[],
                                        modifier: erised::types::TraitBoundModifier::None,
                                    }],
                                    default: None,
                                    synthetic: true,
                                },
                            }],
                            where_predicates: &[],
                        },
                        header: erised::types::Header {
                            const_: false,
                            unsafe_: false,
                            async_: false,
                            abi: erised::types::Abi::Rust,
                        },
                        has_body: true,
                    })
                }),
                erised::types::Identifiable::Item(|| {
                    erised::types::Item::Function(erised::types::Function {
                        name: "pretty_print",
                        meta: erised::types::ItemMeta {
                            krate: || erised::types::ExternalCrate {
                                name: "erised_tests",
                                html_root_url: None,
                            },
                            span: Some(erised::types::Span {
                                filename: "src/lib.rs",
                                begin: (24usize, 0usize),
                                end: (60usize, 1usize),
                            }),
                            visibility: erised::types::Visibility::Public,
                            docs: Some("Use `rustfmt` to pretty-print the tokens."),
                            attrs: &[],
                            deprecation: None,
                        },
                        decl: erised::types::FnDecl {
                            inputs: &[erised::types::FnInput {
                                pat: "tokens",
                                ty: erised::types::Type::ImplTrait(&[
                                    erised::types::GenericBound::TraitBound {
                                        trait_: erised::types::Path {
                                            name: "ToTokens",
                                            prefix: "",
                                            target: erised::types::Identifiable::Summary(|| {
                                                erised::types::ItemSummary {
                                                    krate: || erised::types::ExternalCrate {
                                                        name: "quote",
                                                        html_root_url: Some(
                                                            "https://docs.rs/quote/1.0.26/",
                                                        ),
                                                    },
                                                    path: &["quote", "to_tokens", "ToTokens"],
                                                    kind: erised::types::ItemKind::Trait,
                                                }
                                            }),
                                            args: Some(|| {
                                                erised::types::GenericArgs::AngleBracketed {
                                                    args: &[],
                                                    bindings: &[],
                                                }
                                            }),
                                        },
                                        generic_params: &[],
                                        modifier: erised::types::TraitBoundModifier::None,
                                    },
                                ]),
                            }],
                            output: Some(erised::types::Type::ResolvedPath(erised::types::Path {
                                name: "Result",
                                prefix: "",
                                target: erised::types::Identifiable::Summary(|| {
                                    erised::types::ItemSummary {
                                        krate: || erised::types::ExternalCrate {
                                            name: "core",
                                            html_root_url: Some(
                                                "https://doc.rust-lang.org/nightly/",
                                            ),
                                        },
                                        path: &["core", "result", "Result"],
                                        kind: erised::types::ItemKind::Enum,
                                    }
                                }),
                                args: Some(|| erised::types::GenericArgs::AngleBracketed {
                                    args: &[
                                        erised::types::GenericArg::Type(
                                            erised::types::Type::ResolvedPath(
                                                erised::types::Path {
                                                    name: "String",
                                                    prefix: "",
                                                    target: erised::types::Identifiable::Summary(
                                                        || erised::types::ItemSummary {
                                                            krate: || {
                                                                erised :: types :: ExternalCrate { name : "alloc" , html_root_url : Some ("https://doc.rust-lang.org/nightly/") }
                                                            },
                                                            path: &["alloc", "string", "String"],
                                                            kind: erised::types::ItemKind::Struct,
                                                        },
                                                    ),
                                                    args: Some(|| {
                                                        erised::types::GenericArgs::AngleBracketed {
                                                            args: &[],
                                                            bindings: &[],
                                                        }
                                                    }),
                                                },
                                            ),
                                        ),
                                        erised::types::GenericArg::Type(
                                            erised::types::Type::ResolvedPath(
                                                erised::types::Path {
                                                    name: "Error",
                                                    prefix: "",
                                                    target: erised::types::Identifiable::Summary(
                                                        || erised::types::ItemSummary {
                                                            krate: || {
                                                                erised :: types :: ExternalCrate { name : "anyhow" , html_root_url : Some ("https://docs.rs/anyhow/1.0.70/") }
                                                            },
                                                            path: &["anyhow", "Error"],
                                                            kind: erised::types::ItemKind::Struct,
                                                        },
                                                    ),
                                                    args: Some(|| {
                                                        erised::types::GenericArgs::AngleBracketed {
                                                            args: &[],
                                                            bindings: &[],
                                                        }
                                                    }),
                                                },
                                            ),
                                        ),
                                    ],
                                    bindings: &[],
                                }),
                            })),
                            c_variadic: false,
                        },
                        generics: erised::types::Generics {
                            params: &[erised::types::GenericParamDef {
                                name: "impl ToTokens",
                                kind: erised::types::GenericParamDefKind::Type {
                                    bounds: &[erised::types::GenericBound::TraitBound {
                                        trait_: erised::types::Path {
                                            name: "ToTokens",
                                            prefix: "",
                                            target: erised::types::Identifiable::Summary(|| {
                                                erised::types::ItemSummary {
                                                    krate: || erised::types::ExternalCrate {
                                                        name: "quote",
                                                        html_root_url: Some(
                                                            "https://docs.rs/quote/1.0.26/",
                                                        ),
                                                    },
                                                    path: &["quote", "to_tokens", "ToTokens"],
                                                    kind: erised::types::ItemKind::Trait,
                                                }
                                            }),
                                            args: Some(|| {
                                                erised::types::GenericArgs::AngleBracketed {
                                                    args: &[],
                                                    bindings: &[],
                                                }
                                            }),
                                        },
                                        generic_params: &[],
                                        modifier: erised::types::TraitBoundModifier::None,
                                    }],
                                    default: None,
                                    synthetic: true,
                                },
                            }],
                            where_predicates: &[],
                        },
                        header: erised::types::Header {
                            const_: false,
                            unsafe_: false,
                            async_: false,
                            abi: erised::types::Abi::Rust,
                        },
                        has_body: true,
                    })
                }),
                erised::types::Identifiable::Item(|| {
                    erised::types::Item::Import(erised::types::Import {
                        meta: erised::types::ItemMeta {
                            krate: || erised::types::ExternalCrate {
                                name: "erised_tests",
                                html_root_url: None,
                            },
                            span: Some(erised::types::Span {
                                filename: "src/lib.rs",
                                begin: (8usize, 0usize),
                                end: (8usize, 30usize),
                            }),
                            visibility: erised::types::Visibility::Public,
                            docs: None,
                            attrs: &[],
                            deprecation: None,
                        },
                        source: "erised::heap_types",
                        name: "heap_types",
                        target: Some(erised::types::Identifiable::Summary(|| {
                            erised::types::ItemSummary {
                                krate: || erised::types::ExternalCrate {
                                    name: "erised",
                                    html_root_url: None,
                                },
                                path: &["erised", "heap_types"],
                                kind: erised::types::ItemKind::Module,
                            }
                        })),
                        glob: true,
                    })
                }),
            ],
            is_stripped: false,
        },
        crate_version: Some("0.1.0"),
        all_items: &[],
        summaries: &[],
        external_crates: &[],
    }
}

